fedimint-api/src/db/mem_impl.rs

```diff
@@ -85,6 +95,78 @@ impl IDatabase for MemDatabase {
 
         Ok(())
     }
+
+    fn begin_transaction<'a>(&'a self) -> Box<dyn IDatabaseTransaction<'a> + 'a> {
+        Box::new(MemTransaction {
+            operations: Vec::new(),
+            tx_data: Mutex::new(self.data.lock().unwrap().clone()),
+            db: self,
+        })
+    }
+}
+
+impl<'a> IDatabaseTransaction<'a> for MemTransaction<'a> {
+    fn raw_insert_bytes(&mut self, key: &[u8], value: Vec<u8>) -> Result<Option<Vec<u8>>> {
+        let val = self.raw_get_bytes(key);
+        // Insert data from copy so we can read our own writes
+        self.tx_data
+            .lock()
+            .unwrap()
+            .insert(key.to_vec(), value.clone());
+        self.operations
+            .push(DatabaseOperation::Insert(DatabaseInsertOperation {
+                key: key.to_vec(),
+                value: value,
+            }));
+        val
+    }
+
+    fn raw_get_bytes(&self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        Ok(self.tx_data.lock().unwrap().get(key).cloned())
+    }
+
+    fn raw_remove_entry(&mut self, key: &[u8]) -> Result<()> {
+        // Remove data from copy so we can read our own writes
+        self.tx_data.lock().unwrap().remove(&key.to_vec());
+        self.operations
+            .push(DatabaseOperation::Delete(DatabaseDeleteOperation {
+                key: key.to_vec(),
+            }));
+        Ok(())
+    }
+
+    fn raw_find_by_prefix(&self, key_prefix: &[u8]) -> PrefixIter<'_> {
+        let mut data = self
+            .tx_data
+            .lock()
+            .unwrap()
+            .range::<Vec<u8>, _>((key_prefix.to_vec())..)
+            .take_while(|(key, _)| key.starts_with(key_prefix))
+            .map(|(key, value)| (key.clone(), value.clone()))
+            .collect::<Vec<_>>();
+        data.reverse();
+
+        Box::new(MemDbIter { data })
+    }
+
+    fn commit_tx(self: Box<Self>) -> Result<()> {
```

Comment:
How does this deal with write isolation and isolation levels and what are our requirements?

Basically:

* transaction A starts (gets a copy of data)
* transaction B starts (gets a copy of data)
* transaction A deletes a key X - it works, in the copy of the data
* transaction B deletes a key X - it works, in the copy of the data
* transaction A commits, it works, X gets deleted
* transaction B commits, it should fail as X is deleted and there was a conflict


Seems like rocksdb is implementing a snapshot level transaction isolation? https://www.sobyte.net/post/2022-01/rocksdb-tx/

So to implement it we would need to check which keys were modified since the tx started?

Honestly, I would consider not implementing in-memory database, and just using an ephemeral (temporary, random named file) real database instance for tests. In my experience, database integration is so important and easy to isolate that it's better to just not mock it, take a tiny performance hit and run more tests against the real thing.

Reply:
We chatted offline but I'll respond here for transparency. Yes the above description is exactly right, rocksdb Optimistic Concurrency has snapshot isolation and allows for transactions to read their own writes. The above example with the deletion of the key is how rocksdb would operate.

I very much agree that re-implementing snapshot isolation for an in-memory key value store is not worth our effort. I just didn't want to regress the existing database implementations (sled and in-memory), so essentially I added some hacky code that allows those implementations to operate on a snapshot, but they won't do conflict resolution during commit (because as you point out, that would require MVCC). Since @elsirion mentioned that we currently shouldn't have any write-write conflicts, I think this is ok for now.

I agree that the proper fix is to use rocksdb for the mock tests as well as the real tests and just remove the in-memory and sled support altogether.